Modules/posixmodule.c

```diff
@@ -6915,7 +6915,8 @@ posix_fdopen(PyObject *self, PyObject *args)
         PyMem_FREE(mode);
         return posix_error();
     }
-#if defined(HAVE_FSTAT) && defined(S_IFDIR) && defined(EISDIR)
+#if !defined(MS_WINDOWS) && defined(HAVE_FSTAT) && defined(S_IFDIR) && \
```

Comment:
It's unlikely but still possible that `fd` is a directory. For example, say the directory was opened with `CreateFile` (via ctypes, PyWin32, etc) and the file descriptor was opened with `msvcrt.open_osfhandle`. Granted the `fstat` check has never worked on Windows, but we could actually make it work rather than skip it:

```C
BY_HANDLE_FILE_INFORMATION info;
HANDLE hfile = (HANDLE)_get_osfhandle(fd);
if (GetFileType(hfile) == FILE_TYPE_DISK && 
    GetFileInformationByHandle(hfile, &info) && 
    info.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) {
    ...
```

Reply:
Yes, I considered that `fd` can refer to a directory, but I wasn't sure whether we should bother to fix `dircheck` in 2.7. I mentioned in the bug report that if we do want to fix `dircheck`, we may use 3.x approach, which is the same that you propose (in `_Py_fstat_noraise`). If I get more feedback suggesting to fix `dircheck`, I'll update the patch. Thank you!

Reply:
This suggestion was only for this inline check in `os.fdopen`. I don't think `dircheck` in fileobject.c should even be called on Windows.